modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/IgniteTableImpl.java

```diff
@@ -52,81 +60,117 @@
 import org.apache.ignite.internal.schema.row.RowAssembler;
 import org.apache.ignite.internal.sql.engine.exec.ExecutionContext;
 import org.apache.ignite.internal.sql.engine.exec.RowHandler;
+import org.apache.ignite.internal.sql.engine.exec.TxAttributes;
+import org.apache.ignite.internal.sql.engine.exec.UpdateableTable;
 import org.apache.ignite.internal.sql.engine.exec.exp.RexImpTable;
 import org.apache.ignite.internal.sql.engine.metadata.ColocationGroup;
 import org.apache.ignite.internal.sql.engine.metadata.NodeWithTerm;
 import org.apache.ignite.internal.sql.engine.prepare.MappingQueryContext;
 import org.apache.ignite.internal.sql.engine.rel.logical.IgniteLogicalIndexScan;
 import org.apache.ignite.internal.sql.engine.rel.logical.IgniteLogicalTableScan;
 import org.apache.ignite.internal.sql.engine.schema.IgniteIndex.Type;
-import org.apache.ignite.internal.sql.engine.schema.ModifyRow.Operation;
 import org.apache.ignite.internal.sql.engine.trait.IgniteDistribution;
 import org.apache.ignite.internal.sql.engine.trait.TraitUtils;
 import org.apache.ignite.internal.sql.engine.type.IgniteTypeFactory;
 import org.apache.ignite.internal.sql.engine.util.Commons;
+import org.apache.ignite.internal.sql.engine.util.HashFunctionFactory.RowHashFunction;
 import org.apache.ignite.internal.sql.engine.util.TypeUtils;
 import org.apache.ignite.internal.storage.MvPartitionStorage;
 import org.apache.ignite.internal.table.InternalTable;
+import org.apache.ignite.internal.table.distributed.TableMessagesFactory;
+import org.apache.ignite.internal.table.distributed.replicator.TablePartitionId;
+import org.apache.ignite.internal.table.distributed.replicator.action.RequestType;
+import org.apache.ignite.internal.util.IgniteUtils;
+import org.apache.ignite.lang.ErrorGroups;
+import org.apache.ignite.sql.SqlException;
 import org.jetbrains.annotations.Nullable;
 
 /**
  * Ignite table implementation.
  */
-public class IgniteTableImpl extends AbstractTable implements InternalIgniteTable {
+public class IgniteTableImpl extends AbstractTable implements IgniteTable, UpdateableTable {
+    private static final IgniteLogger LOG = Loggers.forClass(IgniteTableImpl.class);
+
+    private static final TableMessagesFactory MESSAGES_FACTORY = new TableMessagesFactory();
+
     private final TableDescriptor desc;
 
     private final int ver;
 
     private final InternalTable table;
+    private final HybridClock clock;
+    private final ReplicaService replicaService;
 
     private final SchemaRegistry schemaRegistry;
 
     public final SchemaDescriptor schemaDescriptor;
 
     private final Statistic statistic;
 
-    private Map<String, IgniteIndex> indexes = new HashMap<>();
+    private final Map<String, IgniteIndex> indexes = new HashMap<>();
 
     private final List<ColumnDescriptor> columnsOrderedByPhysSchema;
 
+    private final int[] deleteRowHashFields;
+    private final int[] upsertRowHashFields;
```

Comment:
Seems, you can save hashFunctions here instead of caching functions in execution context.

Reply:
Unfortunately, it is not possible, since hashFunction is generified, and table implementation is row format agnostic for the moment being. I filed separate issue to deal with this, because current patch is already too big
https://issues.apache.org/jira/browse/IGNITE-18900